x/group/types/msgs_test.go

```diff
@@ -0,0 +1,350 @@
+	"github.com/cosmos/cosmos-sdk/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	proto "github.com/gogo/protobuf/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestMsgCreateGroupValidation(t *testing.T) {
+	_, _, myAddr := testdata.KeyTestPubAddr()
+	_, _, myOtherAddr := testdata.KeyTestPubAddr()
+
+	specs := map[string]struct {
+		src    MsgCreateGroupRequest
+		expErr bool
+	}{
+		"all good with minimum fields set": {
+			src: MsgCreateGroupRequest{Admin: myAddr},
+		},
+		"all good with a member": {
+			src: MsgCreateGroupRequest{
+				Admin: myAddr,
+				Members: []Member{
+					{Address: myAddr, Power: types.NewDec(1)},
+				},
+			},
+		},
+		"all good with multiple members": {
+			src: MsgCreateGroupRequest{
+				Admin: myAddr,
+				Members: []Member{
+					{Address: myAddr, Power: types.NewDec(1)},
+					{Address: myOtherAddr, Power: types.NewDec(2)},
+				},
+			},
+		},
+		"admin required": {
+			src:    MsgCreateGroupRequest{},
+			expErr: true,
+		},
+		"valid admin required": {
+			src: MsgCreateGroupRequest{
+				Admin: []byte("invalid-address"),
+			},
+			expErr: true,
+		},
+		"duplicate member addresses not allowed": {
+			src: MsgCreateGroupRequest{
+				Admin: myAddr,
+				Members: []Member{
+					{Address: myAddr, Power: types.NewDec(1)},
+					{Address: myAddr, Power: types.NewDec(2)},
+				},
+			},
+			expErr: true,
+		},
+		"negative member's power not allowed": {
+			src: MsgCreateGroupRequest{
+				Admin: myAddr,
+				Members: []Member{
+					{Address: myAddr, Power: types.NewDec(-1)},
+				},
+			},
+			expErr: true,
+		},
+		"empty member's power not allowed": {
+			src: MsgCreateGroupRequest{
+				Admin:   myAddr,
+				Members: []Member{{Address: myAddr}},
+			},
+			expErr: true,
+		},
+		"zero member's power not allowed": {
+			src: MsgCreateGroupRequest{
+				Admin:   myAddr,
+				Members: []Member{{Address: myAddr, Power: sdk.ZeroDec()}},
+			},
+			expErr: true,
+		},
+		"member address required": {
+			src: MsgCreateGroupRequest{
+				Admin: myAddr,
+				Members: []Member{
+					{Power: types.NewDec(1)},
+				},
+			},
+			expErr: true,
+		},
+		"valid member address required": {
+			src: MsgCreateGroupRequest{
+				Admin: myAddr,
+				Members: []Member{
+					{Address: []byte("invalid-address"), Power: types.NewDec(1)},
+				},
+			},
+			expErr: true,
+		},
+	}
+	for msg, spec := range specs {
+		t.Run(msg, func(t *testing.T) {
+			err := spec.src.ValidateBasic()
+			if spec.expErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestMsgCreateGroupSigner(t *testing.T) {
+	_, _, myAddr := testdata.KeyTestPubAddr()
+	assert.Equal(t, []sdk.AccAddress{myAddr}, MsgCreateGroupRequest{Admin: myAddr}.GetSigners())
+}
+
+func TestMsgCreateGroupAccount(t *testing.T) {
+	_, _, myAddr := testdata.KeyTestPubAddr()
+
+	specs := map[string]struct {
+		admin     sdk.AccAddress
+		group     GroupID
+		threshold sdk.Dec
```

Comment:
We don't want to use `sdk.Dec`. Instead, in Regen we want to use `apd.Decimal`. (see ecocredits module)
ref: https://github.com/cosmos/cosmos-sdk/issues/7773

Reply:
Same for weight